<a href="https://colab.research.google.com/github/svfrolov/YP-project/blob/main/%E2%84%9613_%D0%9F%D1%80%D0%BE%D0%B5%D0%BA%D1%82_%D0%B4%D0%BB%D1%8F_%D0%92%D0%B8%D0%BA%D0%B8%D1%88%D0%BE%D0%BF.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Скопирую-в-другой-столбец" data-toc-modified-id="Скопирую-в-другой-столбец-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Скопирую в другой столбец</a></span></li><li><span><a href="#Удаляю-символы,-числа,-привожу-к-нижнему-регистру" data-toc-modified-id="Удаляю-символы,-числа,-привожу-к-нижнему-регистру-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Удаляю символы, числа, привожу к нижнему регистру</a></span></li><li><span><a href="#Разобъю-на-токены" data-toc-modified-id="Разобъю-на-токены-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Разобъю на токены</a></span></li><li><span><a href="#Удаляю-пустые-строки" data-toc-modified-id="Удаляю-пустые-строки-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Удаляю пустые строки</a></span></li><li><span><a href="#Удаляю-стоп-слова-(слова-,-несуществующие-значения)" data-toc-modified-id="Удаляю-стоп-слова-(слова-,-несуществующие-значения)-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Удаляю стоп-слова (слова , несуществующие значения)</a></span></li><li><span><a href="#Лемматизация-текста" data-toc-modified-id="Лемматизация-текста-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Лемматизация текста</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#Разделяю-на-обучающую-и-тестовую" data-toc-modified-id="Разделяю-на-обучающую-и-тестовую-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Разделяю на обучающую и тестовую</a></span></li><li><span><a href="#Разделяю-признаки-в-выборках" data-toc-modified-id="Разделяю-признаки-в-выборках-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Разделяю признаки в выборках</a></span></li><li><span><a href="#Функцици" data-toc-modified-id="Функцици-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Функцици</a></span><ul class="toc-item"><li><span><a href="#функция-обучения-модели" data-toc-modified-id="функция-обучения-модели-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>функция обучения модели</a></span></li></ul></li><li><span><a href="#Модели" data-toc-modified-id="Модели-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Модели</a></span><ul class="toc-item"><li><span><a href="#Decision-Tree-Classifier-10-деревьев" data-toc-modified-id="Decision-Tree-Classifier-10-деревьев-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>Decision Tree Classifier 10 деревьев</a></span></li><li><span><a href="#CatBoost" data-toc-modified-id="CatBoost-2.4.2"><span class="toc-item-num">2.4.2&nbsp;&nbsp;</span>CatBoost</a></span></li><li><span><a href="#Decision-Tree-Classifier-10-деревьев" data-toc-modified-id="Decision-Tree-Classifier-10-деревьев-2.4.3"><span class="toc-item-num">2.4.3&nbsp;&nbsp;</span>Decision Tree Classifier 10 деревьев</a></span></li><li><span><a href="#Decision-Tree-Classifier-100-деревьев" data-toc-modified-id="Decision-Tree-Classifier-100-деревьев-2.4.4"><span class="toc-item-num">2.4.4&nbsp;&nbsp;</span>Decision Tree Classifier 100 деревьев</a></span></li><li><span><a href="#Decision-Tree-Classifier-300-деревьев" data-toc-modified-id="Decision-Tree-Classifier-300-деревьев-2.4.5"><span class="toc-item-num">2.4.5&nbsp;&nbsp;</span>Decision Tree Classifier 300 деревьев</a></span></li><li><span><a href="#CatBoost" data-toc-modified-id="CatBoost-2.4.6"><span class="toc-item-num">2.4.6&nbsp;&nbsp;</span>CatBoost</a></span></li></ul></li></ul></li><li><span><a href="#Тест" data-toc-modified-id="Тест-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Тест</a></span><ul class="toc-item"><li><span><a href="#Оценка-производительности-лучшей-модели-на-тестовой-выборке" data-toc-modified-id="Оценка-производительности-лучшей-модели-на-тестовой-выборке-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Оценка производительности лучшей модели на тестовой выборке</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Выводы</a></span><ul class="toc-item"><li><span><a href="#.-Подготовка" data-toc-modified-id=".-Подготовка-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>. Подготовка</a></span></li><li><span><a href="#.-Обучение" data-toc-modified-id=".-Обучение-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>. Обучение</a></span></li><li><span><a href="#.-После-обучения-получены-следующие-результаты:" data-toc-modified-id=".-После-обучения-получены-следующие-результаты:-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>. После обучения получены следующие результаты:</a></span></li><li><span><a href="#.-Вывод:" data-toc-modified-id=".-Вывод:-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>. Вывод:</a></span></li></ul></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

# Описание проекта

**Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию.**

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.
Постройте модель со значением метрики качества F1 не меньше 0.75.

# Инструкция для проекта
Решить задачу можно как с помощью BERT, так и без этой нейронки. Если хотите попробовать BERT —
Выполните проект локально. В тренажере тетрадь Jupyter ограничена 4 ГБ оперативной памяти — для проекта с BERT этого может не хватить.
Упомяните BERT в заголовке проекта в первой ячейке:
image
Выполнить проект без BERT можно локально или в нашем тренажёре.
В любом случае алгоритм решения выглядит так:

# Загрузите и подготовьте данные.

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

**Данные находятся в файле `/datasets/toxic_comments.csv.` Скачать датасет.**

**Столбец text в нём содержит текст комментария, а** `toxic` — `целевой признак.`

## Подготовка

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer 
import time
from sklearn.model_selection import GridSearchCV

import warnings
warnings.filterwarnings('ignore')

from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from catboost import CatBoostClassifier
from sklearn.metrics import f1_score

import nltk
from nltk import pos_tag
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords, wordnet
from nltk.stem.wordnet import WordNetLemmatizer
nltk.download('averaged_perceptron_tagger')
nltk.download('punkt')
nltk.download('wordnet')

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /home/jovyan/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package punkt to /home/jovyan/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /home/jovyan/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [ ]:
data = pd.read_csv('/datasets/toxic_comments.csv')

In [ ]:
def general_info(data):
            
    print("Количество дублирующихся строк в датафрейме: {}".format(data.duplicated().sum()))
    print()

    print("Общее количество пропущенных значений:")
    display(data.isnull().sum())
    print()
   
    print("Общая информация о датафрейме:")
    print()
    display(data.info())
    print()
    
    print("Первые 5 строк датафрейма:")
    display(data.head())

In [ ]:
general_info(data)

Количество дублирующихся строк в датафрейме: 0

Общее количество пропущенных значений:


Unnamed: 0    0
text          0
toxic         0
dtype: int64


Общая информация о датафрейме:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Unnamed: 0  159292 non-null  int64 
 1   text        159292 non-null  object
 2   toxic       159292 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 3.6+ MB


None


Первые 5 строк датафрейма:


Unnamed: 0                                               text  toxic
0           0  Explanation\nWhy the edits made under my usern...      0
1           1  D'aww! He matches this background colour I'm s...      0
2           2  Hey man, I'm really not trying to edit war. It...      0
3           3  "\nMore\nI can't make any real suggestions on ...      0
4           4  You, sir, are my hero. Any chance you remember...      0

Смотрю токсичные коментарии

In [ ]:
data.query('toxic == 1').head()

Unnamed: 0                                               text  toxic
6            6       COCKSUCKER BEFORE YOU PISS AROUND ON MY WORK      1
12          12  Hey... what is it..\n@ | talk .\nWhat is it......      1
16          16  Bye! \n\nDon't look, come or think of comming ...      1
42          42  You are gay or antisemmitian? \n\nArchangel WH...      1
43          43           FUCK YOUR FILTHY MOTHER IN THE ASS, DRY!      1

### Скопирую в другой столбец

In [ ]:
data['original'] = data['text'].copy()
data = data.reindex(columns=['toxic', 'original', 'text'])

In [ ]:
stopwords_list = stopwords.words('english')
lemmatizer = nltk.stem.WordNetLemmatizer()

In [ ]:
def get_wordnet_pos(word):
    morphy_tag = {'NN':'n', 'JJ':'a',
                  'VB':'v', 'RB':'r'}

    try:
        return morphy_tag[word[:2]]
    except:
        return 'n' 
    
#    return tag_dict.get(tag, wordnet.NOUN)

### Удаляю символы, числа, привожу к нижнему регистру

In [ ]:
def remove_symbols(text):
    return re.sub(r'[\W\d]\s*', ' ', text).lower()

In [ ]:
data.text = data.text.apply(remove_symbols)

### Разобъю на токены

In [ ]:
def tokenize(text):
    return re.split(' ', text)

In [ ]:
data.text = data.text.apply(tokenize)

### Удаляю пустые строки

In [ ]:
def remove_empty(text):
    return [word for word in text if word!='']

In [ ]:
data.text = data.text.apply(remove_empty)

### Удаляю стоп-слова (слова , несуществующие значения)

In [ ]:
def remove_stopwords(text):
    return [word for word in text if word not in stopwords_list]

In [ ]:
data.text = data.text.apply(remove_stopwords)

### Лемматизация текста

In [ ]:
def lemmatize(tokens):
    return [lemmatizer.lemmatize(word.lower(), pos=get_wordnet_pos(tag)) for word, tag in pos_tag(tokens)]

In [ ]:
data.text.isna().sum()

0

In [ ]:
print(type(data.text[0]))

<class 'list'>


In [ ]:
data['lemmatized'] = data['text'].apply(lemmatize)

In [ ]:
data.head()

toxic                                           original  \
0      0  Explanation\nWhy the edits made under my usern...   
1      0  D'aww! He matches this background colour I'm s...   
2      0  Hey man, I'm really not trying to edit war. It...   
3      0  "\nMore\nI can't make any real suggestions on ...   
4      0  You, sir, are my hero. Any chance you remember...   

                                                text  \
0  [explanation, edits, made, username, hardcore,...   
1  [aww, matches, background, colour, seemingly, ...   
2  [hey, man, really, trying, edit, war, guy, con...   
3  [make, real, suggestions, improvement, wondere...   
4                [sir, hero, chance, remember, page]   

                                          lemmatized  
0  [explanation, edits, make, username, hardcore,...  
1  [aww, match, background, colour, seemingly, st...  
2  [hey, man, really, try, edit, war, guy, consta...  
3  [make, real, suggestion, improvement, wonder, ...  
4                [sir, hero, chance, remember, page]

## Обучение

### Разделяю на обучающую и тестовую

In [ ]:
train, test = train_test_split(data, test_size=0.3, random_state=12345)

TF-IDF

In [ ]:
count_tf_idf = TfidfVectorizer()

### Разделяю признаки в выборках

In [ ]:
features_train = train['text'].astype('U')
target_train = train['toxic']
    
features_test = test['text'].astype('U')
target_test = test['toxic']

In [ ]:
features_train = count_tf_idf.fit_transform(features_train)
features_test = count_tf_idf.transform(features_test)

### Функцици

#### функция обучения модели

In [ ]:
models = []

### Модели

In [ ]:
model = LogisticRegression(class_weight='balanced', C=3)

start = time.time()
score_list = cross_val_score(model, features_train, target_train, cv=5, scoring='f1')
end = time.time()
fit_time_lr = end-start
f1_score_lr = score_list.mean()

models.append({'model': model, 'f1_score': f1_score_lr, 'fit_time': fit_time_lr})

In [ ]:
print('LogisticRegression')
print('Время обучения модели:', fit_time_lr)
print('F1-score на валидационной выборке:', f1_score_lr)
print()

LogisticRegression
Время обучения модели: 234.3902223110199
F1-score на валидационной выборке: 0.7607935781783048



#### Decision Tree Classifier 10 деревьев

In [ ]:
model = DecisionTreeClassifier(random_state=12345, max_depth=10)

start = time.time()
score_list = cross_val_score(model, features_train, target_train, cv=5, scoring='f1')
end = time.time()
fit_time_dt = end-start
f1_score_dt = score_list.mean()

models.append({'model': model, 'f1_score': f1_score_dt, 'fit_time': fit_time_dt})

In [ ]:
print('DecisionTreeClassifier')
print('Время обучения модели:', fit_time_dt)
print('F1-score на валидационной выборке:', f1_score_dt)
print()

DecisionTreeClassifier
Время обучения модели: 53.574182748794556
F1-score на валидационной выборке: 0.5802727245348842



#### CatBoost

In [ ]:
model = CatBoostClassifier(random_seed=12345, logging_level='Silent', n_estimators=80)

start = time.time()
score_list = cross_val_score(model, features_train, target_train, cv=5, scoring='f1')
end = time.time()
fit_time_cb = end-start
f1_score_cb = score_list.mean()

models.append({'model': model, 'f1_score': f1_score_cb, 'fit_time': fit_time_cb})

In [ ]:
print('CatBoostClassifier')
print('Время обучения модели:', fit_time_cb)
print('F1-score на валидационной выборке:', f1_score_cb)
print()

CatBoostClassifier
Время обучения модели: 875.2209339141846
F1-score на валидационной выборке: 0.7259859766186729



In [ ]:
results = pd.DataFrame({'Model': ['LR', 'DT', 'CB'],
                        'Время обучения модели:': [fit_time_lr, fit_time_dt, fit_time_cb],
                        'F1-score:': [f1_score_lr, f1_score_dt, f1_score_cb]})
print(results)

  Model  Время обучения модели:  F1-score:
0    LR              234.390222   0.760794
1    DT               53.574183   0.580273
2    CB              875.220934   0.725986


#### Decision Tree Classifier 100 деревьев

In [ ]:
# model = DecisionTreeClassifier(random_state=12345, max_depth=100)
# models.append(fit_model(model, features_train, target_train))
# model = models[-1]['model']
# predictions = predict_values(model, features_test)['predictions']
# f_score_d100 = f1_score(target_test, predictions)
# fit_time_dt100 = models[-1]['fit_time']
# predict_time_dt100 = predict_values(model, features_test)['prediction_time']

In [ ]:
# print('Время обучения модели:', fit_time_dt100)
# print('Время предсказания:', predict_time_dt100)
# print('F1-score:', f_score_d100)

* Время обучения модели: 54.596882820129395
* Время предсказания: 0.023524999618530273
* F1-score: 0.7299855249972164

#### Decision Tree Classifier 300 деревьев

In [ ]:
# model = DecisionTreeClassifier(random_state=12345, max_depth=300)
# models.append(fit_model(model, features_train, target_train))
# model = models[-1]['model']
# predictions = predict_values(model, features_test)['predictions']
# f_score_dt300 = f1_score(target_test, predictions)
# fit_time_dt300 = models[-1]['fit_time']
# predict_time_dt300 = predict_values(model, features_test)['prediction_time']

In [ ]:
# print('Время обучения модели:', fit_time_dt300)
# print('Время предсказания:', predict_time_dt300)
# print('F1-score:', f_score_dt300)

* Время обучения модели: 144.5338273048401
* Время предсказания: 0.04953408241271973
* F1-score: 0.7315514854648066

## Тест

### Оценка производительности лучшей модели на тестовой выборке

In [ ]:
#C=3
model = LogisticRegression(class_weight='balanced', C=3)
start = time.time()
model.fit(features_train, target_train)
end = time.time()

predictions = model.predict(features_test)
f1_score_ = f1_score(target_test, predictions)

In [ ]:
print('Лучшая модель:', 'LogisticRegression')
print('Время обучения модели:', end - start, 'сек.')
print('F1-score на тестовой выборке с помощью LogisticRegression:', f1_score_)

Лучшая модель: LogisticRegression
Время обучения модели: 54.76954889297485 сек.
F1-score на тестовой выборке с помощью LogisticRegression: 0.7644212876814291


## Выводы

### . Подготовка

* Произведена загрузка данных и их подготовка для обучения моделей.

* Представленный датафрейм содержит текстовые комментарии и оценку, является ли комментарий токсичным

* Проведена предобработка текста:

* Удалены символы и числа
* Тексты приведены к нижнему регистру
* Применена токенизация
* Удалены стоп-слова
* Применена лемматизация
* Датафрейм разделил на обучающую и тестовую выборки в соотношении , применил векторизацию TF-IDF

### . Обучение

**Выбраны следующие модели машинного обучения:**

* LogisticRegression
* DecisionTreeClassifier
* CatBoostClassifier

### . После обучения получены следующие результаты:


1. самая высокая метрика F1:  `LogisticRegression` - при быстром обучении, `F-1 = 0.76079`

**за ней следует**


2. `CatBoostClassifier` но медленно обучается, `F-1 = 0.725986`


3.  `DecisionTreeClassifier` F1 на 10 деревьях, очень медленно обучается. `F-1 = 0.58027`

### . Вывод:

1. Самая лучшая модель  `LogisticRegression` на тестовой выборке показала - `F1-score: 0.7644`